### PR TITLE
Begin Profiling Rework to move towards Modularity

### PR DESCRIPTION
**Pull request by maiadegraaf**
This PR introduces a new pragma `custom_profiling_settings` which accepts a path to a JSON file.  The file can be used to turn on or off profiling settings when the profiling mode is set to JSON (`PRAGMA enable_profiling = 'json';`).  If the pragma is not set, all profiling information will be included.

Currently, the settings are as follows:

- **CPU Timing**: The total CPU time spent by that operator and its children.
- **Operator Cardinality**: The number of rows returned by that operator
- **Operator Timing**: The time spent by that operator
- **Extra Info**: Any other information collected during the operation

An example of a custom profiling settings JSON file:
```sql
PRAGMA custom_profiling_settings='/path/to/settings.json'
```
```json
{
  "cpu_time":"true",
  "operator_cardinality":"true",
}
```

----
The plan is to expand this feature to be able to include more settings in the future, as well as a new function in the C API that returns an object with the profiling information.

### CODE REVIEW

File: src/main/settings/settings.cpp [hunk 1]
```diff
@@ -552,6 +552,84 @@ Value EnableProfilingSetting::GetSetting(ClientContext &context) {
...
+	config.profiler_settings.ResetValues();
+}
+
+void CustomProfilingSettings::ResetLocal(ClientContext &context) {
+	auto &config = ClientConfig::GetConfig(context);
+	config.profiler_settings.ResetMetrics();
+	config.profiler_settings.ResetValues();
+}
+
+Value CustomProfilingSettings::GetSetting(ClientContext &context) {
```
Comment (taniabogatsch) [inline src/main/settings/settings.cpp:619]: Let's come back to this in a future PR, as we may want to be able to enable/disable each metric separately and only return the file path here.

File: src/common/tree_renderer.cpp [hunk 2]
```diff
@@ -498,13 +498,20 @@ string TreeRenderer::ExtractExpressionsRecursive(ExpressionInfo &state) {
 }
 
 unique_ptr<RenderTreeNode> TreeRenderer::CreateNode(const QueryProfiler::TreeNode &op) {
```
Comment (taniabogatsch) [inline src/common/tree_renderer.cpp:500]: My editor is showing this function as `unused`. Can we potentially remove it altogether if it is a relic?

```cpp
	unique_ptr<RenderTreeNode> CreateNode(const QueryProfiler::TreeNode &op);
	unique_ptr<RenderTreeNode> CreateNode(const PipelineRenderNode &op);
```

File: src/main/tree_node_settings.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	metrics.clear();
+	metrics = default_metrics;
+}
+
+void TreeNodeSettings::ResetValues() {
+	values.cpu_time = 0;
+	values.extra_info = "";
+	values.operator_cardinality = 0;
+	values.operator_timing = 0;
+}
```
Comment (taniabogatsch) [inline src/main/tree_node_settings.cpp:25]: `values = Metrics()`.

File: src/main/tree_node_settings.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+bool TreeNodeSettings::SettingEnabled(const TreeNodeSettingsType setting) const {
+	if (setting == TreeNodeSettingsType::OPERATOR_TIMING && SettingEnabled(TreeNodeSettingsType::CPU_TIME)) {
+		return true;
+	}
+	return metrics.find(setting) != metrics.end();
```
Comment (taniabogatsch) [inline src/main/tree_node_settings.cpp:31]: Maybe early out if we find it in the map?

File: src/parallel/thread_context.cpp [hunk 5]
```diff
@@ -4,7 +4,8 @@
 
 namespace duckdb {
 
-ThreadContext::ThreadContext(ClientContext &context) : profiler(QueryProfiler::Get(context).IsEnabled()) {
+ThreadContext::ThreadContext(ClientContext &context)
+    : profiler(QueryProfiler::Get(context).IsEnabled(), ClientConfig::GetConfig(context).profiler_settings) {
```
Comment (taniabogatsch) [inline src/parallel/thread_context.cpp:8]: Let's add another constructor or change the existing one.
```cpp
DUCKDB_API explicit OperatorProfiler(bool enabled);
```
becomes
```cpp
DUCKDB_API explicit OperatorProfiler(ClientContext &context);
```

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Hi @maiadegraaf, thanks for the PR, really nice work! I've added the comments that we discussed together. ✌️

File: src/common/tree_renderer.cpp (hunk 2 above)
Comment (maiadegraaf) [reply src/common/tree_renderer.cpp:500]: It doesn't compile without them...

Comment (taniabogatsch) [reply src/common/tree_renderer.cpp:500]: Then we keep them haha.

**Comment by maiadegraaf** [comment]
Had to open a new PR due to some git problems -> #11101

**Pull request closed without merging**